qaoa.py

```diff
@@ -130,3 +146,204 @@ def expectationValue_MaxCut(data,G):
+    cons = []
+    for factor in range(len(bounds)):
+        lower, upper = bounds[factor]
+        l = {'type': 'ineq',
+             'fun': lambda x, lb=lower, i=factor: x[i] - lb}
+        u = {'type': 'ineq',
+             'fun': lambda x, ub=upper, i=factor: ub - x[i]}
+        cons.append(l)
+        cons.append(u)
+    return cons
+
+def optimize_random(K,G, depth=1, weighted=False):
+    """
+    :param K: # Random initializations (RIs)
+    :return: Array of best params (on the format where the gammas and betas are intertwined),
+    the corresponding best energy value, and the average energy value for all the RIs
+    """
+    rec = -np.inf
+    avg_list = np.zeros(K)
+    rec_params = 42 # For debug purposes
+    for i in range(K):
+        init_params = random_init(depth, weighted)
+        cons = get_constaints_for_COBYLA(depth, weighted)
+        sol = minimize(objective_function, x0=init_params, method='COBYLA', args=(G), constraints=cons)
+        params = sol.x
+        qc = createCircuit_MaxCut(params, G, depth)
+        temp_res_data = execute(qc, backend, shots=NUM_SHOTS).result().results
+        E = expectationValue_MaxCut(temp_res_data, G)[0]
+        avg_list[i] = E
+        if E>rec:
+            rec = E
+            rec_params = params
+    return rec_params, rec, np.average(avg_list)
+
+def scale_p(K, G, depth=3, weighted=False):
+    """
+    :return: arrays of the p_values used, the corresponding array for the energy from the optimal
+         energy config, and the average energy (for all the RIs at each p value)
+    """
+    H_list = np.zeros(depth)
+    avg_list = np.zeros(depth)
+    p_list = np.arange(1, depth + 1, 1)
+    for d in range(1, depth + 1):
+        temp, H_list[d-1], avg_list[d-1] = optimize_random(K, G, d, weighted)
+    return p_list, H_list, avg_list
+
+
+
+def INTERP_init(params_prev_step):
+    """
+    Takes the optimal parameters at level p as input and returns the optimal inital guess for
+    the optimal paramteres at level p+1. Uses the INTERP formula from the paper by Zhou et. al
+    :param params_prev_step: optimal parameters at level p
+    :return:
+    """
+    p = len(params_prev_step)
+    params_out_list = np.zeros(p+1)
+    params_out_list[0] = params_prev_step[0]
+    for i in range(2, p + 1):
+        # Next line is clunky, but written this way to accommodate the 1-indexing in the paper
+        params_out_list[i - 1] = (i - 1) / p * params_prev_step[i-2] + (p - i + 1) / p * params_prev_step[i-1]
+    params_out_list[p] = params_prev_step[p-1]
+    return params_out_list
+
+def optimize_INTERP(K, G, depth, weighted=False):
+    """
+    Optimizes the params using the INTERP heuristic
+    :return: Array of the optimal parameters, and the correponding energy value
+    """
+    rec = -np.inf
+    rec_params = 42 # Debug purpose
+    for i in range(K):
+        init_params = np.zeros(2) # NB! This is not motivated from the paper, should be checked further
```

Comment:
what is not motivated?

Reply:
It is not specified from the paper what the initial value should be at level p=1. I will remove the comment in the code.